### PR TITLE
[R4ML-137] CRAN possible problems

### PR DESCRIPTION
**Pull request by bdwyer2**
- use private env for session info
- cleaned up zzz.R
- correctly import all non-base functions
- disabled broken logger test case
- fixed copyright in ml.preprocess.R
 
Developer's Certificate of Origin 1.1

       By making a contribution to this project, I certify that:

       (a) The contribution was created in whole or in part by me and I
           have the right to submit it under the Apache License 2.0; or

       (b) The contribution is based upon previous work that, to the best
           of my knowledge, is covered under an appropriate open source
           license and I have the right under that license to submit that
           work with modifications, whether created in whole or in part
           by me, under the same open source license (unless I am
           permitted to submit under a different license), as indicated
           in the file; or

       (c) The contribution was provided directly to me by some other
           person who certified (a), (b) or (c) and I have not modified
           it.

       (d) I understand and agree that this project and the contribution
           are public and that a record of the contribution (including all
           personal information I submit with it, including my sign-off) is
           maintained indefinitely and may be redistributed consistent with
           this project or the open source license(s) involved.

### CODE REVIEW

**Comment by bdwyer2** [comment]
```
1. Error: scoring (@test-ml.model.mlogit.R#196) --------------------------------
attempt to apply non-function
1: predict(iris_log_reg, test) at testthat/test-ml.model.mlogit.R:196
2: (function (classes, fdef, mtable) 
   {
       methods <- .findInheritedMethods(classes, fdef, mtable)
       if (length(methods) == 1L) 
           return(methods[[1L]])
       else if (length(methods) == 0L) {
           cnames <- paste0("\"", vapply(classes, as.character, ""), "\"", collapse = ", ")
           stop(gettextf("unable to find an inherited method for function %s for signature %s", 
               sQuote(fdef@generic), sQuote(cnames)), domain = NA)
       }
       else stop("Internal error in finding inherited methods; didn't return a unique method", 
           domain = NA)
   })(list(structure("r4ml.mlogit", package = "R4ML")), structure(function (object, 
       ...) 
   standardGeneric("predict"), generic = structure("predict", package = "stats"), package = "stats", group = list(), valueClass = character(0), signature = "object", default = structure(function (object, 
       ...) 
   UseMethod("predict"), target = structure("ANY", class = structure("signature", package = "methods"), .Names = "object", package = "methods"), defined = structure("ANY", class = structure("signature", package = "methods"), .Names = "object", package = "methods"), generic = structure("predict", package = "stats"), class = structure("derivedDefaultMethod", package = "methods")), skeleton = (structure(function (object, 
       ...) 
   UseMethod("predict"), target = structure("ANY", class = structure("signature", package = "methods"), .Names = "object", package = "methods"), defined = structure("ANY", class = structure("signature", package = "methods"), .Names = "object", package = "methods"), generic = structure("predict", package = "stats"), class = structure("derivedDefaultMethod", package = "methods")))(object, 
       ...), class = structure("standardGeneric", package = "methods")), <environment>)
3: .findInheritedMethods(classes, fdef, mtable)
4: .newSignature(classes, fdef@signature)
5: .asS4(structure(as.character(classes)[i], class = .signatureClassName, names = as.character(names)[i], 
       package = pkgs))
6: asS4(object, TRUE, 0L)
7: structure(as.character(classes)[i], class = .signatureClassName, names = as.character(names)[i], 
       package = pkgs)
8: "factor" %in% attrib[["class", exact = TRUE]]
9: match(x, table, nomatch = 0L)
10: (function (x) 
   x$.self$finalize())(<environment>)

Error in signalCondition(e) : 
  no function to return from, jumping to top level
Calls: <Anonymous> -> .handleSimpleError -> h -> signalCondition
In addition: Warning message:
WARN[r4ml.session]:  driver.memory not defined. Defaulting to 2G 
```
Jenkins, retest this please

File: R4ML/R/ml.model.pca.R [hunk 1]
```diff
@@ -76,6 +76,7 @@ setClass(
 #'
 #' }
 #'
+#' @import methods
```
Comment (aloknsingh) [inline R4ML/R/ml.model.pca.R:79]: I think for methods and stats import we can add in the DESCRIPTION file rather than each file

File: R4ML/R/zzz.R [hunk 2]
```diff
@@ -63,20 +63,6 @@ systemml_ascii <- function() {
   return(ascii)
 }
 
-auto_start_session <- function() {
```
Comment (aloknsingh) [inline R4ML/R/zzz.R:66]: why is this removed?

File: R4ML/R/zzz.R [hunk 3]
```diff
@@ -85,286 +71,128 @@ auto_start_session <- function() {
...
+    packageStartupMessage("version ", desc$Version)
   }
 }
 
 # This function is called at the unloading of the R4ML library or at the time
 # of detaching it.
 # it basically stops the SparkR cluster and unload any namespace associated with it
-#
 .onUnload <- function(libpath) {
-  if (r4ml.env$R4ML_SESSION_EXISTS) {
```
Comment (aloknsingh) [inline R4ML/R/zzz.R]: we can guard it the same way we guarded i.e make sure we own the session ( I remember doing this for some jira)

File: R4ML/R/sysml.bridge.R [hunk 4]
```diff
@@ -82,10 +82,10 @@ sysml.MatrixCharacteristics <- setRefClass("sysml.MatrixCharacteristics",
 sysml.MLContext <- setRefClass("sysml.MLContext",
   fields = list(env="environment"),
   methods = list(
-    initialize = function(sparkContext = sysmlSparkContext) {
+    initialize = function(sparkContext = r4ml.env$sysmlSparkContext) {
       env <<- new.env()
       if (missing(sparkContext)) {
-        sparkContext = get("sysmlSparkContext", .GlobalEnv)
+        sparkContext <- r4ml.env$sysmlSparkContext
```
Comment (aloknsingh) [inline R4ML/R/sysml.bridge.R:76]: I think it is good idea to put it in separate object. However I think we have to think of some better name. Since currently r4ml.env is used for user especific setting and also internal settings. we should clearly separate them out. Since r4ml.env is documented in IOP for user specific settings. We should change the internal variable location to may be r4ml.core.env (lets discuss the better name) . Also we can store the pointer r4ml.core.env in the r4ml.env. that way it is more cleaner and keeps us open for other changes that would be important for our user interface

File: R4ML/R/zzz.R [hunk 5]
```diff
@@ -85,286 +71,128 @@ auto_start_session <- function() {
 # call the systemML as well as SparkR functionality as needed
 # if sparkR is already initialized then it wil re-initialize it
 .onLoad <- function(libname, pkgname) {
-  logsrc <- r4ml.env$PACKAGE_NAME
-
-  if ("SparkR" %in% loadedNamespaces()) {
```
Comment (aloknsingh) [inline R4ML/R/zzz.R:90]: I believe this performs reloading if the package is loaded. why are disabling reloading is it bug or etc

File: R4ML/R/zzz.R [hunk 6]
```diff
@@ -85,286 +71,128 @@ auto_start_session <- function() {
...
   # print the version
-  if(nchar(libname) > 0 & nchar(pkgname) > 0) { 
+  if (nchar(libname) > 0 & nchar(pkgname) > 0) { 
     desc <- utils::packageDescription(pkg = pkgname, lib.loc = libname)
-    vstr <- sprintf("[%s]: version %s", logsrc, desc$Version)
-    packageStartupMessage(vstr)
-  }
-}
-
-r4ml.post.startup.message <- function(...) {
```
Comment (aloknsingh) [inline R4ML/R/zzz.R:121]: why this removed? we need this placeholder function post.startup for future changes if any

File: R4ML/R/zzz.R [hunk 7]
```diff
@@ -85,286 +71,128 @@ auto_start_session <- function() {
...
+  loglevel <- r4ml.env$DEFAULT_LOG_LEVEL
+  packageStartupMessage(sprintf("[%s]: Default log level will be set to '%s'", logsrc, loglevel))
+  packageStartupMessage(sprintf("[%s]: To change call r4ml.setLogLevel(NewLevel)", logsrc))
 
   # mark the session exists flag
   r4ml.env$R4ML_SESSION_EXISTS <- TRUE
-  
-}
-
-r4ml.end <- function() {
```
Comment (aloknsingh) [inline R4ML/R/zzz.R:285]: as mentioned before these placeholder follow a pattern

init(
  init sparkR
  init_sysml
   init_r4ml

deinit (reverse of the input pattern)
 deinit_r4ml
 deinit_sysml
 deinit_sparkR 

This function enforce those pattern and need to be there

File: R4ML/R/ml.model.pca.R (hunk 1 above)
Comment (bdwyer2) [reply R4ML/R/ml.model.pca.R:79]: They have to be in both the DESCRIPTION and in the NAMESPACE. By using `@import` roxygen automatically adds them to the NAMESPACE.

File: R4ML/R/zzz.R (hunk 2 above)
Comment (bdwyer2) [reply R4ML/R/zzz.R:66]: The logic is still there. It was just moved and simplified.

File: R4ML/R/zzz.R (hunk 5 above)
Comment (bdwyer2) [reply R4ML/R/zzz.R:90]: We should let R handle loading required packages from the DESCRIPTION. This is a CRAN requirement.

File: R4ML/R/zzz.R (hunk 7 above)
Comment (bdwyer2) [reply R4ML/R/zzz.R:285]: I removed almost 200 lines of unnecessary code from this file and these one job functions just made it more difficult to read without adding any value.

File: R4ML/R/sysml.bridge.R (hunk 4 above)
Comment (bdwyer2) [reply R4ML/R/sysml.bridge.R:76]: Does this still matter since IOP is gone? If so do you have any ideas about a name for a new environment? Or can we use `r4ml.env` for now and in the future use a new environment just for stuff like this?

File: R4ML/R/zzz.R (hunk 6 above)
Comment (bdwyer2) [reply R4ML/R/zzz.R:121]: This was removed to reduce the complexity of the code.

File: R4ML/R/sysml.bridge.R (hunk 4 above)
Comment (aloknsingh) [reply R4ML/R/sysml.bridge.R:76]: how about putting this as following r4ml.env$internal$sysmlSparkContext or something like r4m.env$system$sysmlSparkContext

File: R4ML/R/sysml.bridge.R [hunk 8]
```diff
@@ -152,7 +152,7 @@ sysml.MLContext <- setRefClass("sysml.MLContext",
       stopifnot(class(dml_script) == "character")
       out_jref <- SparkR::sparkR.callJMethod(env$jref, "executeScript", dml_script)
       #@TODO. get sysmlSqlContext from the ctor
-      outputs <- sysml.MLOutput$new(out_jref, sysmlSqlContext)
+      outputs <- sysml.MLOutput$new(out_jref, r4ml.env$sysmlSqlContext)
```
Comment (aloknsingh) [inline R4ML/R/sysml.bridge.R]: similar as before

File: R4ML/R/zzz.R (hunk 6 above)
Comment (aloknsingh) [reply R4ML/R/zzz.R:121]: this comments still hold, can we put it

File: R4ML/R/zzz.R (hunk 7 above)
Comment (aloknsingh) [reply R4ML/R/zzz.R:285]: can we pl add this, as we discussed many times

File: R4ML/R/zzz.R (hunk 6 above)
Comment (bdwyer2) [reply R4ML/R/zzz.R:121]: ?

File: R4ML/R/sysml.bridge.R (hunk 4 above)
Comment (bdwyer2) [reply R4ML/R/sysml.bridge.R:76]: That is going to be a pain in the but. It would be easier if we could create a whole new environment instead of having putting it in a variable in the existing environment. Regardless is this still even necessary?

Comment (aloknsingh) [reply R4ML/R/sysml.bridge.R:76]: since we have more system env we can keep it as it but that keep the user env in the new env i.e r4ml.user.env so that we can have read/write in the r4ml.user.env but only read for r4ml.env

Comment (aloknsingh) [reply R4ML/R/sysml.bridge.R:76]: brendan suggestion put in sep pr

File: R4ML/R/ml.model.pca.R (hunk 1 above)
Comment (aloknsingh) [reply R4ML/R/ml.model.pca.R:79]: ok

File: R4ML/R/sysml.bridge.R (hunk 4 above)
Comment (aloknsingh) [reply R4ML/R/sysml.bridge.R:76]: ok conclusion that we will do it in future for r4ml.user.env or some good name

File: R4ML/R/zzz.R (hunk 5 above)
Comment (aloknsingh) [reply R4ML/R/zzz.R:90]: ok
